Changes

jdarwood007 opened the pull request:
Revamped the ForumSessionProvider and tested on both SMF 2.0 and 2.1.
Added backwards compatibility with the old Auth_SMF.php plugin.

Signed-off-by: jdarwood007 <unmonitored+github@sleepycode.com>

jdarwood007 committed:
Revamped the ForumSessionProvider and tested on both SMF 2.0 and 2.1.
Added backwards compatibility with the old Auth_SMF.php plugin.

Signed-off-by: jdarwood007 <unmonitored+github@sleepycode.com>

Vekseid commented on ForumSessionProvider/ForumSessionProvider.php:
```diff
@@ -18,284 +18,851 @@
+	}
+
+	/*
+	 *	Combatiblity layer for Auth_SMF.php extension settings.
```
I think you mean 'compatibility' rather than combat. Unless you want to slay your old code. >_>

Vekseid commented on ForumSessionProvider/ForumSessionProvider.php:
```diff
@@ -18,284 +18,851 @@
+			$this->FSSettings[$key] = !empty($$key) ? $$key : null;
+	}
+
+	public function __construct(array $params = array())
```
While it may seem like a minor quibble, I'd prefer __construct() to be the first method.

Vekseid commented on ForumSessionProvider/ForumSessionProvider.php:
```diff
@@ -18,284 +18,851 @@
-        if (!in_array(1, $GLOBALS['wgFSPAdminGroups']))
-            $GLOBALS['wgFSPAdminGroups'][] = 1;
+		// Set our software up.
+		$this->wgFSPSoftware = !empty($wgSMFLogin) ? 'smf2.0' : (!empty($wgFSPSoftware) ? $wgFSPSoftware : null);
```
Is this left over from having this as a class variable or something?

jdarwood007 replied:
This is entirely for compatiblity with the old settings from Auth_SMF plugin.  If that setting exists, we enforced SMF 2.0 (as the plugin was technically not for 2.1) and continue on.

Vekseid replied:
No, I mean wgFSPSoftware isn't a defined class property in your revision. So I'm not sure why you're referencing it as such rather than just using the $wgFSPSoftware global.

jdarwood007 replied:
Ahh yea, I see what you mean.  Will add that to a protected.

Vekseid commented on ForumSessionProvider/ForumSessionProvider.php:
```diff
@@ -18,284 +18,851 @@
 	protected $id = 0;
-	protected $password = '';
-	protected $prefix;
+	protected $FScookie = array();
```
MediaWiki 1.27 requires PHP 5.5.9 or 5.6, shorthand array syntax is available and should be used.

jdarwood007 replied:
SMF has always spelled it out, since we've existed before [] was a thing.  I've continued to always spell it out just out of habit.

Vekseid commented on ForumSessionProvider/ForumSessionProvider.php:
```diff
@@ -18,284 +18,851 @@
+	*/
+	public static function redirect($action, $do_return = false)
+	{
+		global $wgScriptPath, $wgFSPSoftware, $boardurl, $image_proxy_secret;
```
I don't see image_proxy_secret being assigned as global in the read settings function? Is this working correctly?

jdarwood007 replied:
Its a a static function.  To get the settings we setup earlier, I would need to make this a singleton call and a public function to read out the setting.  Was just easier to do it this way.

It is working.  I've put a PR in place for 2.1.  Technically that PR could backport to 2.0 and work the same.

Vekseid replied:
To be more specific, is this actually seeing the $image_proxy_secret that we believe it is?

jdarwood007 replied:
That variable is from SMF Settings.php, not from MW.  But yes it is what we believe it is.

Vekseid commented on ForumSessionProvider/ForumSessionProvider.php:
```diff
@@ -18,284 +18,851 @@
+			case 'elk1.1':
+				return $this->FScookie['password'] === hash('sha256', $this->FSuser['passwd'] . $this->FSuser['password_salt']);
+				break;
+			case 'smf2.0':
```
I don't see any reason for this to be split out.

jdarwood007 replied:
There is a reason for it.  More on that later.

Vekseid replied:
So, what's the story with making these into functions?

jdarwood007 replied:
Nothing I can post yet.  If you want to email me directly I could share some details.

Vekseid replied:
@jdarwood007 I emailed you at sleepy @ simplemachines. Let me know if there's a better one to contact you at.

Vekseid commented on ForumSessionProvider/ForumSessionProvider.php:
```diff
@@ -18,284 +18,851 @@
+			case 'smf2.0':
+				return $this->checkPasswordSMF20();
+				break;
+			case 'smf2.1':
```
Unless some drastic modification is still in the cards for SMF 2.1, I don't see why this is split out either.

Vekseid commented on ForumSessionProvider/ForumSessionProvider.php:
```diff
@@ -18,284 +18,851 @@
+	{
+		global $wgRequest;
+
+		// !!! TODO: Caching this?  Can't use $_SESSION ANYMORE.
```
Does MediaWiki provide hooks to its cache handlers? IMO we should use those rather than trying to cook up our own, if possible.

jdarwood007 replied:
I was looking into that, but couldn't find any good simple examples of the caching.  I find MediaWiki code hard to read as its overly complex.  I couldn't find a straight forward example of the caching process.

But it would be ideal to cache this.  Every time the user hits the page, it decodes the cookie, connects to the FS database and issues a query against it.  Seems to me if we can cache that, we could issue fewer queries.  Especially if we can cache it in a users session.  So while they are browsing the wiki, it maybe only hits the FS database 1-2 times in a 15 minute window.

Vekseid replied:
Not disagreeing with the need to cache, though it's a bit of a lesser concern for me. Should probably straight make an issue for caching this and bring discussion over.

Vekseid commented on ForumSessionProvider/ForumSessionProvider.php:
```diff
@@ -18,284 +18,851 @@
+			'LoginDeniedGroups' => 'wgFSPDenyGroups',
+			'LoginAllowedGroups' => 'wgFSPAllowGroups',
+			'AdminGroups' => 'wgFSPAdminGroups',
+			'SuperAdminGroups' => 'wgFSPSuperGroups',
```
You appear to have forgotten interface groups here.

Vekseid commented on ForumSessionProvider/ForumSessionProvider.php:
```diff
@@ -18,284 +18,851 @@
+
+		// Only SMF does this.
+		if ($this->FSSettings['NameStyle'] == 'smf')
+			$user->setOption('forum_member_id', $this->FSuser['id_member']);
```
What purpose does this serve? It can become inaccurate as members change names.

jdarwood007 replied:
The purpose is to attempt to prevent account takeovers.  Since its possible in SMF to change the account name, this would essentially open up the account on the wiki side to be taken over by another member.  This logic is to stop that.  Although there is no clear way of fixing this from the wiki side other than deleting the account.

At one point I did intend to have logic if it couldn't find a member by the name, to instead go search to find the forum member id, then force a name change.

Vekseid replied:
Elkarte doesn't work any differently, I just don't allow account name changes willy nilly on my forums.

There's a lot to unpack with this and this isn't a properly complete solution if you want to address this. The situation depends on software (currently the same for everyone) and to prevent such takeovers it needs to be checked in the validation stage like everything else. 

Should remove this for now, and make a separate PR with a more complete solution.

Vekseid replied:
Err, sorry if it wasn't clear. There is no point in gating this behind the name style.

jdarwood007 replied:
If I remove the gating for now, I think we can come back around to improving this.  The idea is sound, just yes I haven't gotten around to fully implanting it.

With the ID as well we can reverse link their wiki profile to their forum profile.  We could also maybe allow a wiki admin to unlink the profile as well, allowing it to be linked by another.

The real problem is how both handle allow changing usernames.  It becomes a problem on both sides.  This was my first step in that direction to allow us to ensure accounts can be changed and relinked correctly.

Vekseid commented on ForumSessionProvider/ForumSessionProvider.php:
```diff
@@ -18,284 +18,851 @@
+	// !!! Add checks here for banning
+	private function isBanned()
+	{
+		switch ($this->wgFSPSoftware)
```
I intentionally left this out. Elkarte's banning schema is largely the same as SMF's is for now.

I'd prefer an option to check for bans. The switch can remain but can safely include everyone for now.

jdarwood007 replied:
Sounds reasonable.  I don't know elks db to know what is compatible.

Vekseid replied:
The only difference is how IPs are handled, slightly.

I'm planning on a more comprehensive IP replacement which SMF may want to adopt but that's a whole 'nother discussion.

Vekseid commented on ForumSessionProvider/ForumSessionProvider.php:
```diff
@@ -18,284 +18,851 @@
+		);
+
+		// As of now, we don't suport anything other than UTF8 with SMF.
+		$smcFunc['db_query']('set_character_set', '
```
I hacked this for Elliquiy anyway. Please do not assume this.

jdarwood007 replied:
Suggestions?  SMF does have a character set we could check for.  Its largely ignored in 2.1 since we migrated to full UTF-8.

Vekseid replied:
I wrote this before I wrote the DB layer comment, sorry. Should have removed this.

Vekseid commented on ForumSessionProvider/ForumSessionProvider.php:
```diff
@@ -18,284 +18,851 @@
+	 *
+	 * {@source}
+	 */
+	private function FSDBConnect()
```
Are there really enough queries to justify incorporating SMF's DB engine, as opposed to just supporting Postgres plain? I fear this will just lead to a mess of trying to support both forums' own DB engines rather than just handling it directly.

jdarwood007 replied:
Either way is fine.  With this syntax we also do get the benefit of using the database layer and its protections.

Vekseid replied:
For... two queries, both of which are hunting for an integer, I think we can trust ourselves to cast that.  Best just to load PDO and switch to that where needed.

jdarwood007 replied:
I will try to get time this weekend to change it up to a PDO setup.

Vekseid commented on ForumSessionProvider/ForumSessionProvider.php:
```diff
@@ -18,284 +18,851 @@
+	/*
+	 * Inject the SMF AuthManager hook to override stuff we no longer need them to change
+	*/
+	private function AuthManagerHook()
```
There is nothing in here unique to SMF (or any forum period), we should just hook it the proper way and block email/password changes, etc. raw.

jdarwood007 replied:
Sounds fair, let me do some testing on that.

jdarwood007 replied:
I did also do this so if we disable the extension, it becomes editable again.  Instead of having to disable it elsewhere as well.

Vekseid replied:
Woot. : )

Vekseid commented on ForumSessionProvider/ForumSessionProvider.php:
```diff
@@ -18,284 +18,851 @@
+	}
+}
+
+class AuthManager_SMF extends TemporaryPasswordPrimaryAuthenticationProvider
```
See comment above - we should incorporate these hooks directly.

Vekseid reviewed (commented):
First review, before testing. 

I am especially leery about incorporating database layers. It feels excessive for so few queries.

jdarwood007 committed:
Some changes
Figured out a simple caching

Signed-off-by: jdarwood007 <unmonitored+github@sleepycode.com>

jdarwood007 committed:
Generic Names

Signed-off-by: jdarwood007 <unmonitored+github@sleepycode.com>

Vekseid commented on ForumSessionProvider/extension.json:
```diff
@@ -37,7 +45,8 @@
   "AutoloadClasses": {
-    "ForumSessionProvider": "ForumSessionProvider.php"
+    "ForumSessionProvider": "ForumSessionProvider.php",
+    "AuthManager_SMF": "ForumSessionProvider.php"
```
If you want to keep this a separate classes, can just as well add a second file under the name.

Also worry that SMF branding may confuse people in thinking it's more specific than it is.

Vekseid commented on ForumSessionProvider/extension.json:
```diff
@@ -46,7 +46,7 @@
   "AutoloadClasses": {
     "ForumSessionProvider": "ForumSessionProvider.php",
-    "AuthManager_SMF": "ForumSessionProvider.php"
+    "AuthManager_ForumSoftware": "ForumSessionProvider.php"
```
For naming consistences, I'd suggest ForumAuthManager, and again make it a separate file.

Vekseid commented on ForumSessionProvider/extension.json:
```diff
@@ -30,8 +30,8 @@
   	"primaryauth": {
-		"AuthManager_SMF": {
-		  "class": "AuthManager_SMF",
+		"AuthManager_ForumSoftware": {
```
Are you sure there are no hooks to add for this here?

jdarwood007 replied:
MW code is difficult to read but I didn't see any when poking around.  The best way I could see was to create the class that uses the standard auth as its base.

jdarwood007 committed:
Use PDO instead of any specific database driver.
Generic names

Signed-off-by: jdarwood007 <unmonitored+github@sleepycode.com>

jdarwood007 commented:
I haven't wrote many pdos, so yea.

Don't have a postgresql anymore but it should work from what I read.

jdarwood007 committed:
Protect this var.

Signed-off-by: jdarwood007 <unmonitored+github@sleepycode.com>

jdarwood007 committed:
Allow forum id for all.

jdarwood007 committed:
2.0.16 changes

jdarwood007 committed:
Remove the functions

Vekseid commented:
Alright, I'm going to put your version through some live testing and see how it goes.

Thank you. : )

Vekseid commented on ForumSessionProvider/ForumSessionProvider.php:
```diff
@@ -14,288 +14,753 @@
+	{
+		global $wgSMFPath, $wgFSDebug, $wgFSPDenyGroups, $wgFSPAllowGroups, $wgFSPAdminGroups, $wgFSPSuperGroups, $wgFSPSpecialGroups, $wgFSPInterfaceGroups, $wgFSPEnableBanCheck;
+
+		$this->FSSettings['path'] = isset($wgSMFPath) ? $wgSMFPath : '../forum';
```
This clobbers the original and documented $wgFSPPath. Keep the SMF setting in your compat.

Vekseid commented on ForumSessionProvider/ForumSessionProvider.php:
```diff
@@ -14,288 +14,753 @@
+
+		// Simple caching?
+		try {
+			$cache = MediaWikiServices::getInstance()->getLocalServerObjectCache();
```
getLocalServerObjectCache does not exist in MediaWiki 1.27

Vekseid reviewed (changes requested):
These two at a minimum need fixing, will test again when these are addressed.

jdarwood007 committed:
Minor bugs

Vekseid commented on ForumSessionProvider/ForumSessionProvider.php:
```diff
@@ -14,288 +14,759 @@
+			'SuperAdminGroups' => 'wgFSPSuperGroups',
+			'InterfaceGroups' => 'wgFSPInterfaceGroups',
+			'SpecialGroups' => 'wgFSPSpecialGroups',
+			'NameStyle' => 'wgFSPNameStyle',
```
This isn't getting properly read. Nor is it defaulting properly when it fails to read it.

Vekseid commented:
Few more comments.

In any case, this isn't working for me, in SMF 2.0 or in Elkarte. I need to dig down deeper into this, or maybe ask you to split up the changes so we can test and apply things in a saner fashion.

jdarwood007 committed:
Updates to logging, and adjustments to globals

Merge branch 'master' of https://github.com/Vekseid/ForumSessionProvider

# Conflicts:
#	ForumSessionProvider/ForumAuthManager.php
#	ForumSessionProvider/ForumSessionProvider.php
#	ForumSessionProvider/extension.json

jdarwood007 commented:
Your welcome to close this PR if you feel its not needed anymore.  Did a merge but kept my changes and fixed a few things that I saw could be wrong.

Vekseid commented:
It is getting rather hard for me to make sense of. Some things got clobbered, I can tell.

If there's anything specific you see, please feel free to submit specific PRs or open issues for them. Your caching changes at least still need to make it in, though I can do that once I've got a proper environment for it.

In any case, thank you for your help. : )

jdarwood007 commented:
The caching is one change.  I'm testing on 1.31 as its the current LTS, if your testing against something older you should change that: https://www.mediawiki.org/wiki/Version_lifecycle
I need to test against 1.35 as well soon.

The only other big difference here is I'm putting everything into FSSettings and using that rather than globals or a bunch of local variable calls.  Just cleaned up a lot of stuff.
